### PR TITLE
Automatically close the garage when parked at home

### PR DESCRIPTION
**Pull request by dhnoronha**

### CODE REVIEW

**Review by brchri — CHANGES_REQUESTED** [review]
I like this idea and appreciate the PR. I've left a couple of implementation comments. Also, I think auto closing on park at home should be a configurable option, as not everyone may want this. Some people may park and then walk out to the street to get their mail, for example, and don't want the auto close, so it should be added to the config to enable this feature. It should probably be a per-car config item, as some drivers in a household may want this, and others may not.

I'm open to further discussion on this and the other comments I've left.

File: internal/geo/geo.go [hunk 1]
```diff
@@ -113,6 +118,52 @@ func CheckGeofence(config util.ConfigStruct, car *util.Car) {
 	car.GarageDoor.OpLock = false                                     // release garage door's operation lock
 }
 
+// when parked, check if the car is at the home location and close the garage door.
+func VailidateGeofenceAndClose(config util.ConfigStruct, car *util.Car, action string) {
```
[the remark this answers is not in the document]
Comment (dhnoronha) [reply internal/geo/geo.go:122]: Makes sense, Will be removed.

File: internal/geo/geo.go [hunk 2]
```diff
@@ -113,6 +118,52 @@ func CheckGeofence(config util.ConfigStruct, car *util.Car) {
...
+	timeout := 40
```
[the remark this answers is not in the document]
Comment (dhnoronha) [reply internal/geo/geo.go:123]: Switching to closing the garage from within the CheckGeofence function based on a car configuration should avoid the need for this timeout.

**Comment by DjAnu** [comment]
Wow this is awesome, I had been googling around for past few days to somehow Auto open/close Garage using Tesla location. I have read hundreds of forums , tried using HA, Homebridge and what not, none of those are perfect solutions. Code you guys are working on is way more perfect. One of the major issue I have with using Geolocation is that my home at a location where for circular geofences (which is the only option for iOS and all other devices / HA I use) I cross the geofence a couple of times everytime I am leaving or coming home (see the screenshot below). I love the idea of having polygon fences. Hope at a later time I can somehow use it for other automations.......This is a perfect project to work on this weekend. I am not a programmer though so I hope I'll be able to figure out the integration process you have outlined here and if not may be I can reach out to you guys for help.....
Screenshot 2023-09-15 152033

**Comment by brchri** [comment]
> Wow this is awesome, I had been googling around for past few days to somehow Auto open/close Garage using Tesla location. I have read hundreds of forums , tried using HA, Homebridge and what not, none of those are perfect solutions. Code you guys are working on is way more perfect. One of the major issue I have with using Geolocation is that my home at a location where for circular geofences (which is the only option for iOS and all other devices / HA I use) I cross the geofence a couple of times everytime I am leaving or coming home (see the screenshot below). I love the idea of having polygon fences. Hope at a later time I can somehow use it for other automations.......This is a perfect project to work on this weekend. I am not a programmer though so I hope I'll be able to figure out the integration process you have outlined here and if not may be I can reach out to you guys for help..... Screenshot 2023-09-15 152033

Hey thanks for the love. I agree there are some other good solutions out there but none of them did what I wanted, so I started this project and open sourced it in hopes others might find it useful. If you have any issues getting it going, feel free to open an issue on the repo and I'm happy to help out, and it would help me learn what gaps there are in the documentation to try and make it as self-serve as possible. 😄

**Comment by DjAnu** [comment]
> @DjAnu Hey thanks for the love. I agree there are some other good solutions out there but none of them did what I wanted, so I started this project and open sourced it in hopes others might find it useful. If you have any issues getting it going, feel free to open an issue on the repo and I'm happy to help out, and it would help me learn what gaps there are in the documentation to try and make it as self-serve as possible. 😄

Hey thanks for your reply. I was able to set up telsamate using following instructions : 
https://geekjournal.ch/how-to-install-teslamate-a-great-data-logger-and-visualizer-for-your-tesla/

I also updated the config file and added polygon coordinates. However as I am running docker on my synology NAS, I couldn't figure out how to load and run your image on my docker. Sorry I am not a programmer :-(. I am still searching to see if I can find any tutorial. If you have some time can you please give me some pointers. OR if we can hop on a char or something that would be great. Tx.

**Comment by DjAnu** [comment]
Okay seems like I am getting close. I was able to load and run the docker using the instructions below: 
https://mariushosting.com/how-to-install-firefox-on-your-synology-nas/
 I ran the user defined scripts for both loading the docker & to get the MyQ serial number and was able to run it successfully.

I am still running into 2 issues.
1. app isn't able to connect to MQTT broker
2023/09/16 18:16:48 could not connect to mqtt broker: network Error : dial tcp 143.244.220.150:1883: i/o timeout
2. Where do I get teslamate_car_id

**Comment by brchri** [comment]
@DjAnu Congrats on getting that far.

1. The message `dial tcp 143.244.220.150:1883: i/o timeout` looks like it's trying to connect to a public IP address, which I highly doubt is correct. You'd want it to be the IP address of your synology most likely, if that's where you're hosting Teslamate. So in your Tesla-YouQ `config.yml` file, you'd set `mqtt_host: <ip of synology>`.
2. The short answer is that if this is the first car you've connected, your `teslamate_car_id` almost certainly `1`. The long answer is there are a few ways to get the teslamate car id, either by looking at the database or exploring the MQTT broker. You can use a tool like [this](https://github.com/thomasnordquist/MQTT-Explorer) to connect to your MQTT broker and look around, and as you expand the tree, you'll see `<your synology IP>/cars/<your teslamate car id>`. Using this tool will also help you validate that your MQTT broker from Teslamate is exposed properly so Tesla-YouQ can access it.

Hope that helps.

**Comment by DjAnu** [comment]
Thanks for your comment. I was able to figure out car-id from teslamate logs but I am still not able to find MQTT broker IP. I have tried synology IP, container name etc but still no go :-(. I downloaded the MQTT explorer portable but I can't seem to connect using IP or hostname using that too. Not sure how can I find MQTT server. Teslamate seem to connect fine though. Any pointer here ? Screenshot 2023-09-16 200421

**Comment by brchri** [comment]
@DjAnu Based on that log it looks like the host should be `192.168.0.11`. I took a look at the guide you linked, and it looks like the [docker-compose.yml example](https://geekjournal.ch/wp-content/uploads/2022/01/docker-compose-example.txt) they provided has commented out the port exposure for the MQTT broker. If you also have those commented out in your config file, you'll want to remove the comments so the ports will actually be exposed. That would cause your problem.

So instead of this:

```
mosquitto:
    image: eclipse-mosquitto:2
    restart: always
    command: mosquitto -c /mosquitto-no-auth.conf
    # ports:
    #   - 1883:1883
    volumes:
      - mosquitto-conf:/mosquitto/config
      - mosquitto-data:/mosquitto/data
```

You'd have this:

```
mosquitto:
    image: eclipse-mosquitto:2
    restart: always
    command: mosquitto -c /mosquitto-no-auth.conf
    ports:
      - 1883:1883
    volumes:
      - mosquitto-conf:/mosquitto/config
      - mosquitto-data:/mosquitto/data
```

Note I removed the `#` from two lines.

Hope that helps.

**Comment by DjAnu** [comment]
Yes, I had figured that out and had made changes accordingly. Do you use
any channel / app where we can chat a little bit if you don't mind. It
could be reditt or google voice or anything you use. I think I am very
close just can't figure out MQTT connection. Truly appreciate your help.

**Comment by brchri** [comment]
@DjAnu Yeah I'm on reddit, [Fridge9165](https://www.reddit.com/user/Fridge9165/)

**Pull request closed without merging**